enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,369 @@
+
+### Risks and Mitigations
+
+## Design Details
+
+The existing ovn-kubernetes hybrid overlay feature was developed for hybrid
+Windows/Linux clusters. Each ovn-kubernetes node manages an external-to-OVN OVS
+bridge, named br-ext, which acts as the VXLAN source and endpoint for packets
+moving between pods on the node and their cluster-external destination. The
+br-ext SDN switch acts as a transparent gateway and routes traffic towards
+windows nodes.
+
+In the SDN live migration use case, we can enhance this feature to connects the
+nodes managed by different CNI plugins. To minimize the implementation effort,
+and the code maintainability, we will try to reused the whole hybrid overlay and
+only make necessary changes to both CNI plugins.
+
+On the OVN Kubernetes side, all the cross CNI traffic shall follow the same path
+of current hybrid overlay implementation. For OVN Kubernetes, we need to do
+following enhancements:
+
+1. We need to modify OVN-K to allow overlapping between the cluster network and
+   the Hybrid overlay CIDR. So that we can reuse the cluster network in the
+   migration.
+2. We need to modify OVN-K to allow modifying the hybrid overlay CIDR on the
+   fly.
+3. We need to allow `hybrid-overlay-node` running on linux nodes, currently it
+   is designed to only run on windows nodes. 
+
+On the OpenShift SDN side, when a node is converted to OVN Kubernetes,
+it shall be almost transparent to the control-plane. But we need to introduce a
+'migration mode' for OpenShift SDN by:
+
+1. Change ingress NetworkPolicy processing to be based entirely on pod IPs
+   rather than using namespace VNIDs, since packets from ovn nodes will have
+   the VNID 0 set.
+2. To be compatible with Windows node VXLAN implementation, OVN-K hybrid overlay
+   uses the host interface MAC as VXLAN inner MAC. When packets arrive at the
+   br0 of the SDN node, they cannot be forwarded to the pod interface, due to
+   the MAC mismatch. We need to add flows for each pod which swaps the dst MAC
+   to the pod interface mac.
+
+### The Traffic Path
+
+#### Packets going from OpenShift SDN to OVN Kubernetes
+
+On the SDN side, it doesn't need to know if the peer node is a SDN node or a OVN
+node. We reuse the existing VXLAN tunnel rules one the SDN side.
+- Egress NetworkPolicy rules and service proxying happen as normal
+- When the packet reaches table 90, it will hit a “send via vxlan” rule that was
+  generated based on a HostSubnet object.
+
+On the OVN side:
+- OVN accepts the packet via the VXLAN tunnel, ignore the VNID set by SDN and
+  then just routes it normally.
+- Ingress NetworkPolicy processing will happen when the packet reaches the
+  destination pod’s switch port, just like normal.
+  - Our NetworkPolicy rules are all based on IP addresses, not “logical input
+    port”, etc, so it doesn’t matter that the packets came from outside OVN and
+    have no useful OVN metadata.
+
+#### Packets going from OVN Kubernetes to OpenShift SDN
```

Comment:
We'll also have to modify the hybrid overlay code to handle SDN nodes having HostSubnets within the ClusterCIDR; because Windows HO doesn't do that. So I guess ovnkube will have to learn about HostSubnet resources, or maybe the SDN code could annotate its nodes with the right bits so ovnkube can grab them subnets.

Reply:
Instead of letting ovnkube-master read the HostSubnet CR, I plan to run the `hybrid-overlay-node` binary on the SDN nodes. It will be responsible for reading the HostSubnets CR and annotating the node. So it won't affect the current Windows HO logic.

Reply:
oh this is why you need to run hybrid-overlay-node on linux nodes. You should make that clear in my earlier comment that it is for linux nodes running openshift-sdn

Reply:
will do